src/core/print.ml

```diff
@@ -94,83 +134,75 @@ and pp_term : term pp = fun oc t ->
     | Vari(x)     -> pp_tvar oc x
     | Type        -> out oc "TYPE"
     | Kind        -> out oc "KIND"
-    | Symb(s,h)   -> pp_symbol h oc s
+    | Symb(s)     -> pp_symbol oc s
     | Meta(m,e)   -> out oc "%a%a" pp_meta m pp_env e
     | Patt(_,n,e) -> out oc "$%s%a" n pp_env e
     | TEnv(t,e)   -> out oc "$%a%a" pp_term_env t pp_env e
     (* Product and abstraction (only them can be wrapped). *)
-    | Abst(a,t)   ->
+    | Abst(a,b)   ->
         if wrap then out oc "(";
-        let pp_arg oc (x,a) =
-          if !print_domains then out oc "(%a:%a)" pp_tvar x (pp `Func) a
-          else pp_tvar oc x
-        in
-        let (x,t) = Bindlib.unbind t in
-        out oc "λ%a" pp_arg (x,a);
-        let rec pp_absts oc t =
-          match unfold t with
-          | Abst(a,t) -> let (x,t) = Bindlib.unbind t in
-                         out oc " %a%a" pp_arg (x,a) pp_absts t
-          | _         -> out oc ", %a" (pp `Func) t
-        in
-        pp_absts oc t;
+        out oc "λ";
+        let (x,t) = Bindlib.unbind b in
+        if Bindlib.binder_occur b then out oc "%a" pp_tvar x else out oc "_";
+        if !print_domains then out oc ": %a" (pp `Func) a;
+        out oc ", %a" (pp `Func) t;
         if wrap then out oc ")"
     | Prod(a,b)   ->
         if wrap then out oc "(";
-        let pp_arg oc (x,a) = out oc "(%a:%a)" pp_tvar x (pp `Func) a in
-        let (x,c) = Bindlib.unbind b in
+        let (x,t) = Bindlib.unbind b in
         if Bindlib.binder_occur b then
-          begin
-            out oc "Π%a" pp_arg (x,a);
-            let rec pp_prods oc c =
-              match unfold c with
-              | Prod(a,b) when Bindlib.binder_occur b ->
-                  let (x,b) = Bindlib.unbind b in
-                  out oc " %a%a" pp_arg (x,a) pp_prods b
-              | _                                      ->
-                  out oc ", %a" (pp `Func) c
-            in
-            pp_prods oc c
-          end
```

Comment:
Same here, why did you remove the more compact printing?

Reply:
I prefer this one. It has less parentheses. Too much parentheses is bad for readability.

Reply:
It has maybe less parentheses, but it introduces several dependent product symbols.
Do you really prefer `Π a : A, Π b : B, Π c : C, D` over `Π (a : A) (b : B) (c : C), D`? The choice does not seem obvious to me. And we can modify the later to factor consecutive variables of the same type so this would become `Π a : A, Π b : A, Π c : A, B` against `Π (a b c : A), B`, which is a clear win. This would have been a trivial modification from the factored one.

Reply:
Mathematicians do not write parentheses everywhere.  Π a b c : A, B is good as well and can be done in another PR.